Add Dockerfile and publish image to container registries

imiric opened the pull request:
This allows using xk6 via Docker, without having to setup a local Go environment or install xk6.

Unfortunately, file permissions with mounted volumes are [still an issue](https://github.com/moby/moby/issues/7198), without a simple, native fix. [fixuid](https://github.com/boxboat/fixuid) seems to be the modern solution to this, and works well from my tests.

The `grafana/xk6` image isn't published yet, which will be done after this is merged, and we make a new release. I've tested publishing to GHCR in my fork, which [worked fine](https://github.com/imiric/xk6/pkgs/container/xk6).

In the meantime, you can build it with e.g. `docker build -t test/xk6 .`.

Closes #8

imiric commented:
It would probably be good to add some CI tests for the image, right?

I can do that later today.

olegbespalov commented on README.md:
````diff
@@ -9,11 +9,34 @@ It is used heavily by k6 extension developers as well as anyone who wishes to ma
+For example, to build a k6 v0.43.0 binary with the [xk6-kafka](https://github.com/mostafa/xk6-kafka) and [xk6-output-influxdb](https://github.com/grafana/xk6-output-influxdb) extensions, you would run:
+
+```bash
+docker run --rm -it -u "$(id -u):$(id -g)" -v "$PWD:/xk6" grafana/xk6 build v0.43.0 \
````
Non-blocking suggestion

```suggestion
docker run --rm -it -u "$(id -u):$(id -g)" -v "${PWD}:/xk6" grafana/xk6 build v0.43.0 \
```

If I got [this correctly](https://stackoverflow.com/a/41489151) (I don't have windows with docker installed), it should work both on linux/windows power shell

imiric replied:
This _might_ work on PowerShell, but not on cmd.exe, which uses `%cd%` instead of `$PWD`. I'm actually surprised `id` exists on Windows, so those commands should actually work. :crossed_fingers:

olegbespalov reviewed (approved):
I belive the image test in CI step is a requirement for image publishing, so I'm fine to address it in the following PR.

:+1:

imiric commented on Dockerfile:
```diff
@@ -0,0 +1,28 @@
+ARG GO_VERSION=1.20.1
+ARG VARIANT=bullseye
+FROM golang:${GO_VERSION}-${VARIANT} as builder
```
This `builder` stage is currently unused, and I want to change the way xk6 is currently built from `go install` to copying the checked out repo in the container and running `go build` instead. This would make it work with forks, which is currently not possible with `go install`.

mstoykov reviewed (approved):
LGTM :clap: 

Let's see now how many people have really hard time getting the mounting of a local direcotry to work  :sweat_smile:

imiric commented:
I'm not quite finished with this, as I want to invite people on macOS and Windows using Docker Desktop to try this out, and see if it works for them as well. :crossed_fingers: 

And while I tested the Docker release to GHCR in my fork, we also need to push the image to Docker Hub, so I'll need some help from DevOps, or whoever is managing the Grafana org to create the initial package we can push to.

imiric commented:
OK, this should be done now. We confirmed the image works on macOS and Windows, and I updated the instructions.

I also added pushing to Docker Hub. [grafana/xk6](https://hub.docker.com/r/grafana/xk6/) is now available, but I haven't yet tested pushing to it. I added the secrets to the repo, and mostly copied the commands from the k6 workflow, so it _should_ work once this is merged. :crossed_fingers:

olegbespalov reviewed (approved):
:heart:

javaducky reviewed (approved):
LGTM! Confirmed on Apple silicon.

olegbespalov commented on README.md:
````diff
@@ -9,11 +9,72 @@ It is used heavily by k6 extension developers as well as anyone who wishes to ma
+```
+
+For the native Windows binary if you're using cmd.exe:
+```batch
````
```suggestion
```bash
```

imiric replied:
Actually, this is Windows Batch file syntax. I'm not sure if `cmd.exe` or `batch` would work, but `batch` seems to highlight some syntax, so it's good enough. :man_shrugging:
